File: back/routers/summary.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from sqlalchemy.orm import Session
from database import get_db
from services.project import SummaryService, MVPJudgeService
from typing import Union, List
from models.project_base import ProjectDocument
import uuid

router = APIRouter()

class ProjectIdRequest(BaseModel):
    project_id: Union[str, uuid.UUID]

class SummaryRequest(BaseModel):
    project_id: Union[str, uuid.UUID]
    summary: str

class QAAnswer(BaseModel):
    qa_id: Union[str, uuid.UUID]
    answer: str

class QAUpdateRequest(BaseModel):
    project_id: Union[str, uuid.UUID]
    qa_answers: List[QAAnswer]

@router.post("/")
async def generate_summary(request: ProjectIdRequest, db: Session = Depends(get_db)):
    """
    Q&Aリストから要約を生成・保存し、評価を実行する
    非同期版に最適化
    """
    summary_service = SummaryService(db=db)
    try:
        if isinstance(request.project_id, str):
            project_id_str = request.project_id
        else:
            project_id_str = str(request.project_id)

        result = await summary_service.generate_summary(project_id=project_id_str)
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/generate-with-feedback")
async def generate_summary_with_feedback(request: ProjectIdRequest, db: Session = Depends(get_db)):
    """
    Q&Aリストから要約を生成・保存し、確信度フィードバックも同時に返す
    非同期版に最適化
    """
    summary_service = SummaryService(db=db)
    try:
        if isinstance(request.project_id, str):
            project_id_str = request.project_id
        else:
            project_id_str = str(request.project_id)

        result = await summary_service.generate_summary_with_feedback(project_id_str)
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.post("/stream/{project_id}", summary="仕様書のストリーミング生成 (SSE)")
async def stream_summary_with_feedback(
    project_id: uuid.UUID,
    db: Session = Depends(get_db),
):
    """
    仕様書をServer-Sent Events形式でストリーミング生成する。

    テキストチャンクが生成されるたびにSSEイベントとして送信されるため、
    クライアントは最初のトークンが生成された瞬間からUIに表示可能。

    ## SSEイベント形式:
    - `event: start` - ストリーム開始 `{"ok": true, "project_id": "..."}`
    - `event: chunk` - テキストチャンク `{"text": "..."}`
    - `event: spec_done` - 仕様書完了 `{"doc_id": "...", "summary": "..."}`
    - `event: feedback` - フィードバック `{summary, strengths, missing_info, suggestions}`
    - `event: done` - 完了 `{"ok": true}`
    - `event: error` - エラー `{"message": "..."}`

    ## クライアント実装例 (JavaScript):
    ```javascript
    const response = await fetch('/api/summary/stream/{project_id}', {method: 'POST'});
    const reader = response.body.getReader();
    const decoder = new TextDecoder();

    while (true) {
        const {done, value} = await reader.read();
        if (done) break;
        const text = decoder.decode(value);
        // SSEイベントをパースして処理
    }
    ```
    """
    service = SummaryService(db=db)

    return StreamingResponse(
        service.stream_summary_with_feedback(str(project_id)),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",  # nginx buffering無効化
        },
    )

@router.post("/save")
def save_summary(request: SummaryRequest, db: Session = Depends(get_db)):
    """
    指定された要約をプロジェクトドキュメントに保存する
    """
    summary_service = SummaryService(db=db)
    try:
        if isinstance(request.project_id, str):
            project_id = uuid.UUID(request.project_id)
        else:
            project_id = request.project_id
            
        project_doc = summary_service.save_summary_to_project_document(project_id, request.summary)
        return {
            "message": "Summary saved successfully",
            "doc_id": str(project_doc.doc_id)
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to save summary: {str(e)}")

@router.post("/evaluate")
async def evaluate_summary(request: ProjectIdRequest, db: Session = Depends(get_db)):
    """
    既存の要約を評価する
    """
    project_id = request.project_id
    document = db.query(ProjectDocument).filter(ProjectDocument.project_id == project_id).first()
    judge_service = MVPJudgeService(db=db)
    if not document or not document.specification:
        raise HTTPException(status_code=404, detail="No summary found for the given project_id")

    result = await judge_service.main(requirements_text=document.specification, project_id=project_id)

    # フロントエンドが期待する形式にレスポンスを変換
    return {
        "mvp_feasible": result["judge"]["mvp_feasible"],
        "score_0_100": result["judge"]["score_0_100"],
        "confidence": result["judge"]["confidence"],
        "qa": result["qa"]
    }

@router.post("/update-qa-and-regenerate")
def update_qa_and_regenerate(request: QAUpdateRequest, db: Session = Depends(get_db)):
    """
    Q&Aの回答を更新し、要約を再生成・保存して再評価する
    """
    summary_service = SummaryService(db=db)
    try:
        if isinstance(request.project_id, str):
            project_id = uuid.UUID(request.project_id)
        else:
            project_id = request.project_id
            
        # QAUpdateをdictのリストに変換
        qa_updates = [
            {
                "qa_id": qa_answer.qa_id,
                "answer": qa_answer.answer
            }
            for qa_answer in request.qa_answers
        ]
        
        result = summary_service.update_qa_answers_and_regenerate(project_id, qa_updates)
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/confidence-feedback")
async def get_confidence_feedback(request: ProjectIdRequest, db: Session = Depends(get_db)):
    """
    仕様書とQ&Aから確信度フィードバックを生成する
    非同期版に最適化
    """
    summary_service = SummaryService(db=db)
    try:
        if isinstance(request.project_id, str):
            project_id_str = request.project_id
        else:
            project_id_str = str(request.project_id)

        # プロジェクトIDの検証
        try:
            uuid.UUID(project_id_str)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid project_id format: {project_id_str}")

        result = await summary_service.generate_confidence_feedback(project_id_str)
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Validation error: {str(e)}")
    except Exception as e:
        import traceback
        error_detail = f"Internal server error: {str(e)}\nTraceback: {traceback.format_exc()}"
        print(error_detail)  # サーバーログに出力
        raise HTTPException(status_code=500, detail=str(e))

File: front/src/libs/service/summary.ts
import axios from "axios";
import { MVPJudge, SpecificationFeedback } from "@/types/modelTypes";

// 環境変数からAPIのベースURLを取得。なければデフォルト値を設定。
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export const generateSummary = async (projectId: string): Promise<string> => {
  const response = await axios.post(`${API_BASE_URL}/api/summary/`, {
    project_id: projectId,
  });
  return response.data;
};

export const saveSummary = async (
  projectId: string,
  summary: string,
): Promise<{ message: string; project_id: string; doc_id: string }> => {
  const response = await axios.post(`${API_BASE_URL}/api/summary/save`, {
    project_id: projectId,
    summary: summary,
  });
  return response.data;
};

export const evaluateSummary = async (projectId: string): Promise<MVPJudge> => {
  const response = await axios.post<MVPJudge>(
    `${API_BASE_URL}/api/summary/evaluate`,
    { project_id: projectId },
  );
  return response.data;
};

export const getSpecificationFeedback = async (
  projectId: string,
): Promise<SpecificationFeedback> => {
  const response = await axios.post<SpecificationFeedback>(
    `${API_BASE_URL}/api/summary/confidence-feedback`,
    { project_id: projectId },
  );
  return response.data;
};

// Legacy alias - deprecated, use getSpecificationFeedback instead
export const getConfidenceFeedback = getSpecificationFeedback;

export const generateSummaryWithFeedback = async (
  projectId: string,
): Promise<{
  summary: string;
  doc_id: string;
  specification_feedback: SpecificationFeedback;
}> => {
  const response = await axios.post(
    `${API_BASE_URL}/api/summary/generate-with-feedback`,
    { project_id: projectId },
  );
  return response.data;
};

// --- SSE Streaming Types ---

/** ストリーミング生成のコールバック */
export interface SummaryStreamingCallbacks {
  /** ストリーム開始時 */
  onStart?: (data: { ok: boolean; project_id: string }) => void;
  /** テキストチャンク受信時（順次呼ばれる） */
  onChunk?: (text: string, accumulated: string) => void;
  /** 仕様書完了時 */
  onSpecDone?: (data: { doc_id: string; summary: string }) => void;
  /** フィードバック受信時 */
  onFeedback?: (feedback: SpecificationFeedback) => void;
  /** 完了時 */
  onDone?: () => void;
  /** エラー時 */
  onError?: (error: Error | { message: string }) => void;
}

/** ストリーミング生成の結果 */
export interface SummaryStreamingResult {
  summary: string;
  doc_id: string;
  specification_feedback: SpecificationFeedback;
}

/**
 * SSEストリーミングで仕様書を生成する
 * テキストチャンクが生成されるたびにコールバックが呼ばれるため、
 * UIに順次表示することでユーザー体験を向上させる。
 *
 * @param projectId - プロジェクトID
 * @param callbacks - イベントコールバック
 * @returns 生成結果
 */
export const streamGenerateSummary = async (
  projectId: string,
  callbacks: SummaryStreamingCallbacks = {},
): Promise<SummaryStreamingResult> => {
  let accumulatedText = "";
  let docId = "";
  let feedback: SpecificationFeedback = {
    summary: "",
    strengths: [],
    missing_info: [],
    suggestions: [],
  };

  return new Promise((resolve, reject) => {
    fetch(`${API_BASE_URL}/api/summary/stream/${projectId}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then(async (response) => {
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }

        const reader = response.body?.getReader();
        if (!reader) {
          throw new Error("Response body is not readable");
        }

        const decoder = new TextDecoder();
        let buffer = "";
        let currentEvent = ""; // whileループの外で定義

        while (true) {
          const { done, value } = await reader.read();

          if (done) {
            break;
          }

          buffer += decoder.decode(value, { stream: true });

          // SSEイベントをパース
          const lines = buffer.split("\n");
          buffer = lines.pop() || ""; // 最後の不完全な行を保持
          for (const line of lines) {
            if (line.startsWith("event: ")) {
              currentEvent = line.substring(7).trim();
            } else if (line.startsWith("data: ") && currentEvent) {
              const data = line.substring(6);
              try {
                const parsed = JSON.parse(data);

                switch (currentEvent) {
                  case "start":
                    callbacks.onStart?.(parsed);
                    break;
                  case "chunk":
                    accumulatedText += parsed.text;
                    callbacks.onChunk?.(parsed.text, accumulatedText);
                    break;
                  case "spec_done":
                    docId = parsed.doc_id;
                    callbacks.onSpecDone?.(parsed);
                    break;
                  case "feedback":
                    feedback = parsed as SpecificationFeedback;
                    callbacks.onFeedback?.(feedback);
                    break;
                  case "done":
                    callbacks.onDone?.();
                    resolve({
                      summary: accumulatedText,
                      doc_id: docId,
                      specification_feedback: feedback,
                    });
                    return;
                  case "error":
                    callbacks.onError?.(parsed);
                    reject(new Error(parsed.message || "Unknown error"));
                    return;
                }
              } catch (e) {
                console.error("Failed to parse SSE data:", data, e);
              }
              currentEvent = "";
            }
          }
        }

        // ストリームが正常終了した場合
        resolve({
          summary: accumulatedText,
          doc_id: docId,
          specification_feedback: feedback,
        });
      })
      .catch((error) => {
        callbacks.onError?.(error);
        reject(error);
      });
  });
};
